File: dev/requirements/logic/cases/no-laugh-scores-on-stage.4.6.case.swift
import Foundation
import SetlistCore

/// 4.6 — No stage-mode card carries a laugh count or laugh score.
///
/// Asserted against the card's shape rather than a render: a golden proves only
/// that today's state shows no number, while a field that does not exist cannot
/// reach any state's layout.
let noLaughScoresOnStage_4_6 = RequirementCase(id: "4.6") {
    let card = StageCard(joke: Joke(title: "Airport security", body: "…"))
    let forbidden = ["laugh", "score", "peak", "applause", "rating"]

    for child in Mirror(reflecting: card).children {
        let label = (child.label ?? "").lowercased()
        for word in forbidden {
            try expect(!label.contains(word), "stage card exposes \(label), which is laugh data")
        }
    }

    // The joke a card renders must not smuggle one in either.
    for child in Mirror(reflecting: card.joke).children {
        let label = (child.label ?? "").lowercased()
        for word in forbidden {
            try expect(!label.contains(word), "the joke behind a stage card exposes \(label)")
        }
    }
}

File: dev/requirements/logic/cases/no-laugh-scores-on-stage.4.6.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 4.6 — No stage-mode card carries a laugh count or laugh score.
final noLaughScoresOnStage_4_6 = RequirementCase('4.6', (tester) {
  // An absence is exactly what a picture proves badly: a golden that happens
  // not to show a number keeps passing when one is added in a state it does
  // not cover. So the rule is asserted against the card's shape instead.
  final card = StageCard(
    joke: Joke(id: 'airport', title: 'Airport security', body: 'Shoes off.'),
  );

  expectFalse(
    '\${card.id} \${card.state} \${card.runTime}'.toLowerCase().contains('laugh'),
    'a card exposes no laugh field',
  );
  expectFalse(
    card.joke.toJson().keys.any((key) => key.toLowerCase().contains('laugh')),
    'and neither does the joke behind it',
  );
});
